### ISSUE TITLE
3.2.7: Hidden Controls - what happens with optional / on-demand controls

### ISSUE THREAD

**Issue by patrickhlauke** [w3c/wcag#1400]
The SC is fairly clear about "controls needed to progress or complete a process", but what happens on pages that have optional / on-demand processes? For instance, a page that primarily *shows* the user's account credentials, but also allows the user to edit details if they need to. The page's primary purpose is to show the current information, but has an ancilliary functionality to edit things. Would this SC require that these controls be always visible / have an option to make them permanently visible? I guess this would depend on the understanding of what the "process" for that page is, with the extra complication that this may depend on what the user's current intention actually is (which is not something an author can necessarily foresee).

**Comment by patrickhlauke** [comment]
good examples (came up in discussion with @smhigley) as well: imagine something like a treeview that represents a file system. nominally the purpose of it is to navigate through this file system and open folders/files. for each file/folder, there's also additional functionality to delete/edit that file/folder, and a button for context menu with even more options. does this SC make it mandatory for those delete, edit, menu buttons to be visible at all times? it would result in a completely overwhelming, cluttered interface full of trashcan, pencil, three-dot menu icons which will, from a usability perspective, make the experience far worse for all users, including users with cognitive disabilities.

in short, the concept is good, but i think it hinges heavily on the vague definition of "process", on what "needed to progress or complete" is depending on what the user is actually trying to accomplish, etc. and with this vagueness there's a strong chance that this will be misinterpreted/misapplied to the detriment of all users.

**Comment by mraccess77** [comment]
Yes, I've raised the issue of what a process is on the working group calls months ago.  I believe this is very helpful for things like completing a form, checking out, etc. and even for controls that can be edited but don't look interactive but more problematic for places like editors, menus, trees, etc. where space is limited and conventions are present.  While those things (bolding text, deleting a file etc.) could be a process adding all of those icons could be good or bad depending on the user and ultimately and option (or just a single disclosure widget) would be better.    I'd note that in Windows you can often show or hide more of these things in Explorer through settings.   Gmail offers options when you hover over messages - but also allows you to open the message and reach the same features.

**Comment by patrickhlauke** [comment]
Looking closer at the understanding document

> Controls should be visible without user interaction (i.e., persistently visible) when they are needed to progress.

and then the examples

> Web Editor Success: In a web editor, once a page is open for editing, all the possible locations for editing display a pencil icon.

so by implication, is this saying that the "process" on the editor page is to be able to edit? it sounds less like a "process" in the sense of

> series of user actions where each action is required in order to complete an activity

to me, as the user *may* or *may not* **want** to edit each piece of content, or not. a user can proceed/leave the page without even editing a single thing, i'd assume. in this case the "process" is heavily dependent on what the user's actual intent is at that particular moment, something which the author may not know in advance...so the result would be, for a reasonably sized page to edit, potentially dozens/hundred of "pencil" icons littered all over the page (next to each heading, paragraph, bullet point, image, line break, etc).

> Some design approaches hide controls needed to complete tasks and require certain user interactions, such as mouse-over, to display these controls.

do we have any more tangible examples of these "design approaches"? i generally can't think of cases i've encountered where crucial controls that are necessary to move forward in process are hidden/hard to find - since that impacts all users. generally, i'd say hidden controls are more common when particular tasks/options are deemed to be optional/ancilliary/context-specific. and in those cases they're hidden exactly because otherwise they'd completely overwhelm the user with a cluttered mess of controls.

**Comment by patrickhlauke** [comment]
> People with low executive function, impaired memory, and other cognitive and learning disabilities may not be able to find controls needed to progress if they are hidden until focus is placed on them or a pointer hovers over them.

this mentions controls that only appear when focus/hover is on/over *them*. what if the controls appear when focusing on/hovering over a large visible item (so it's not on/over *them* per se, but rather the item they refer to). e.g. focusing on/hovering over an entry in a list to reveal additional controls next to the item (delete/edit/menu). a strict reading of this part of the understanding would suggest they're exempt, since the user doesn't have to remember where those hidden controls are (and risk not finding them again), but remember that focusing/hovering clearly visible items will lead to further options/controls being revealed in context.

**Comment by mraccess77** [comment]
I agree with Patrick here -- the wording isn't right -- basically the wording need to say if you hover or focus a control and additional controls appear... because you won't likely know to hover over a control that isn't on the screen!

**Comment by alastc** [comment]
HI @patrickhlauke,

Thanks for raising this. The [latest version](https://w3c.github.io/wcag/understanding/visible-controls.html) uses a different approach which focuses on the information for that control being available is visible. There is no use of process.

I'll close this as the source of the issue is gone, but please do raise any new issues on the new text, it will be going out for review again.
